File: src/main/java/de/twometer/neko/event/Events.java
package de.twometer.neko.event;

import de.twometer.neko.util.Log;
import org.greenrobot.eventbus.EventBus;

public final class Events {

    public static void init() {
        // Create useful event bus
        EventBus.builder()
                .logNoSubscriberMessages(false)
                .sendNoSubscriberEvent(false)
                .installDefaultEventBus();
        Log.d("Event system initialized");
    }

    public static void register(Object subscriber) {
        EventBus.getDefault().register(subscriber);
    }

    public static void unregister(Object subscriber) {
        EventBus.getDefault().unregister(subscriber);
    }

    public static void post(Object o) {
        EventBus.getDefault().post(o);
    }

}

File: neko-engine/src/main/kotlin/de/twometer/neko/events/Events.kt
package de.twometer.neko.events

import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe

object Events {

    fun setup() {
        EventBus.builder()
            .logNoSubscriberMessages(false)
            .sendNoSubscriberEvent(false)
            .installDefaultEventBus()
    }

    fun register(subscriber: Any) {
        if (!hasSubscriberMethods(subscriber))
            return
        EventBus.getDefault().register(subscriber)
    }

    fun unregister(subscriber: Any) {
        if (!EventBus.getDefault().isRegistered(subscriber))
            return
        EventBus.getDefault().unregister(subscriber)
    }

    fun post(event: Any) = EventBus.getDefault().post(event)

    private fun hasSubscriberMethods(subscriber: Any): Boolean {
        return subscriber.javaClass.methods.any { it.isAnnotationPresent(Subscribe::class.java) }
    }

}

data class ResizeEvent(val width: Int, val height: Int)

data class FocusEvent(val focused: Boolean)

data class CharEvent(val char: Char, val codepoint: Int)

data class KeyEvent(val key: Int, val scancode: Int, val action: Int, val mods: Int)

data class KeyPressEvent(val key: Int)

data class MouseMoveEvent(val xPos: Int, val yPos: Int)

data class MouseClickEvent(val button: Int)

data class MouseButtonEvent(val button: Int, val action: Int, val mods: Int)

data class MouseWheelEvent(val xOff: Int, val yOff: Int)

class RenderDeferredEvent

class RenderForwardEvent

class TickEvent